File: packages/hmssdk_flutter/ios/Classes/Actions/HMSVideoAction.swift
//
//  HMSVideoAction.swift
//  hmssdk_flutter
//
//  Created by govind on 03/02/22.
//

import Foundation
import HMSSDK

class HMSVideoAction {
    static func videoActions(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?, _ swiftHmssdkFlutterPlugin: SwiftHmssdkFlutterPlugin) {
        switch call.method {
        case "switch_video":
            switchVideo(call, result, hmsSDK)

        case "switch_camera":
            switchCamera(result, hmsSDK, swiftHmssdkFlutterPlugin)

        case "is_video_mute":
            isVideoMute(call, result, hmsSDK)

        case "mute_room_video_locally":
            toggleVideoMuteAll(true, result, hmsSDK)

        case "un_mute_room_video_locally":
            toggleVideoMuteAll(false, result, hmsSDK)

        case "toggle_camera_mute_state":
            toggleCameraMuteState(result, hmsSDK, swiftHmssdkFlutterPlugin)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    static private func switchCamera(_ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?, _ swiftHmssdkFlutterPlugin: SwiftHmssdkFlutterPlugin) {
        guard let peer = hmsSDK?.localPeer,
              let videoTrack = peer.videoTrack as? HMSLocalVideoTrack
        else {
            if swiftHmssdkFlutterPlugin.previewForRoleVideoTrack != nil {
                let videoTrack = swiftHmssdkFlutterPlugin.previewForRoleVideoTrack!
                videoTrack.switchCamera()
                result(nil)
                return
            }
            result(HMSErrorExtension.getError("Local Peer not found in \(#function)"))
            return
        }

        videoTrack.switchCamera()

        result(nil)
    }

    static private func switchVideo(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {

        let arguments = call.arguments as! [AnyHashable: Any]

        guard let shouldMute = arguments["is_on"] as? Bool,
              let peer = hmsSDK?.localPeer,
              let video = peer.videoTrack as? HMSLocalVideoTrack else {
            result(false)
            return
        }

        video.setMute(shouldMute)

        result(true)
    }

    static private func toggleCameraMuteState(_ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?, _ swiftHmssdkFlutterPlugin: SwiftHmssdkFlutterPlugin) {

            guard let peer = hmsSDK?.localPeer,
              let video = peer.videoTrack as? HMSLocalVideoTrack else {
                if swiftHmssdkFlutterPlugin.previewForRoleVideoTrack != nil {
                    let videoTrack = swiftHmssdkFlutterPlugin.previewForRoleVideoTrack!
                    videoTrack.setMute(!(videoTrack.isMute()))
                    result(true)
                    return
                }
            result(false)
            return
        }

        video.setMute(!(video.isMute()))

        result(true)
    }

    static private func isVideoMute(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {
        let arguments = call.arguments as! [AnyHashable: Any]

        if let peerID = arguments["peer_id"] as? String, let peer = HMSCommonAction.getPeer(by: peerID, hmsSDK: hmsSDK) {
            if let video = peer.videoTrack {
                result(video.isMute())
                return
            }
        } else {
            if let peer = hmsSDK?.localPeer, let video = peer.videoTrack {
                result(video.isMute())
                return
            }
        }

        result(false)
    }

    static private func toggleVideoMuteAll(_ shouldMute: Bool, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {

        if let localPeer = hmsSDK?.localPeer {
            if let video = localPeer.videoTrack as? HMSLocalVideoTrack {
                video.setMute(shouldMute)
            }
        }

        let room = hmsSDK?.room
        if room != nil {
            let videoTracks = HMSUtilities.getAllVideoTracks(in: room!) as [HMSVideoTrack]?
            videoTracks?.forEach { track in
                if track is HMSRemoteVideoTrack {
                    (track as! HMSRemoteVideoTrack).setPlaybackAllowed(!shouldMute)
                }
            }
        }
        result(nil)
    }
}

File: packages/hmssdk_flutter/android/src/main/kotlin/live/hms/hmssdk_flutter/methods/HMSVideoAction.kt
package live.hms.hmssdk_flutter
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import live.hms.video.media.tracks.*
import live.hms.video.sdk.*
import live.hms.video.sdk.models.*
import live.hms.video.utils.HmsUtilities

class HMSVideoAction {
    companion object {
        fun videoActions(call: MethodCall, result: Result, hmssdk: HMSSDK, hmssdkFlutterPlugin: HmssdkFlutterPlugin?) {
            when (call.method) {
                "switch_video" -> {
                    switchVideo(call, result, hmssdk)
                }

                "switch_camera" -> {
                    switchCamera(result, hmssdk, hmssdkFlutterPlugin)
                }

                "is_video_mute" -> {
                    result.success(isVideoMute(call, hmssdk))
                }

                "mute_room_video_locally" -> {
                    toggleVideoMuteAll(true, result, hmssdk)
                }

                "un_mute_room_video_locally" -> {
                    toggleVideoMuteAll(false, result, hmssdk)
                }

                "toggle_camera_mute_state" -> {
                    toggleCameraMuteState(result, hmssdk, hmssdkFlutterPlugin)
                }

                else -> {
                    result.notImplemented()
                }
            }
        }

        private fun switchVideo(call: MethodCall, result: Result, hmssdk: HMSSDK) {
            val argsIsOn = call.argument<Boolean>("is_on")
            val peer = hmssdk.getLocalPeer()
            val videoTrack = peer?.videoTrack
            if (videoTrack != null) {
                videoTrack.setMute(argsIsOn ?: false)
                result.success(true)
            } else {
                result.success(false)
            }
        }

        private fun toggleCameraMuteState(result: Result, hmssdk: HMSSDK, hmssdkFlutterPlugin: HmssdkFlutterPlugin?) {
            val peer = hmssdk.getLocalPeer()
            val videoTrack = peer?.videoTrack
            if (videoTrack != null) {
                videoTrack.setMute(!(videoTrack.isMute))
                result.success(true)
            } else {
                // /Checking whether preview for role audio track exist or not
                if (hmssdkFlutterPlugin?.previewForRoleVideoTrack != null) {
                    val previewTrack = hmssdkFlutterPlugin.previewForRoleVideoTrack
                    previewTrack?.setMute(!(previewTrack.isMute))
                    result.success(true)
                } else {
                    result.success(false)
                }
            }
        }

        private fun switchCamera(result: Result, hmssdk: HMSSDK, hmssdkFlutterPlugin: HmssdkFlutterPlugin?) {
            val peer = hmssdk.getLocalPeer()
            var videoTrack = peer?.videoTrack
            if (videoTrack == null) {
                // /Checking whether preview for role audio track exist or not
                if (hmssdkFlutterPlugin?.previewForRoleVideoTrack != null) {
                    videoTrack = hmssdkFlutterPlugin.previewForRoleVideoTrack
                } else {
                    return
                }
            }
            CoroutineScope(Dispatchers.Default).launch {
                videoTrack?.switchCamera(onAction = HMSCommonAction.getActionListener(result))
            }
        }

        private fun isVideoMute(call: MethodCall, hmssdk: HMSSDK): Boolean {
            val peerId = call.argument<String>("peer_id")
            if (peerId == "null") {
                return hmssdk.getLocalPeer()?.videoTrack?.isMute ?: true
            }
            val peer = HMSCommonAction.getPeerById(peerId!!, hmssdk)
            return peer?.videoTrack?.isMute ?: true
        }

        private fun toggleVideoMuteAll(shouldMute: Boolean, result: Result, hmssdk: HMSSDK) {
            val room: HMSRoom? = hmssdk.getRoom()
            if (room != null) {
                val videoTracks: List<HMSVideoTrack> = HmsUtilities.getAllVideoTracks(room)
                videoTracks.forEach { it ->
                    if (it is HMSRemoteVideoTrack) {
                        it.isPlaybackAllowed = (!shouldMute)
                    }
                }
                HMSCommonAction.getLocalPeer(hmssdk)!!.videoTrack?.setMute((shouldMute))
            }
        }
    }
}
